exit/test_exit.c

```diff
@@ -0,0 +1,518 @@
+	}
+
+	pid = fork();
+	if (pid < 0) {
+		TEST_ASSERT_LESS_THAN_INT(0, pid);
+	}
+	/* child */
+	if (pid == 0) {
+		/* Exit right away */
+		_exit(EXIT_SUCCESS);
+	}
+	/* parent */
+	else {
+		ret = wait(&status);
+		TEST_ASSERT_EQUAL_INT(pid, ret);
+
+		/* Wait until handler end execution */
+		sleep(1);
+		errno = 0;
+		ret = open(PATH, O_RDWR);
+		TEST_ASSERT_NOT_EQUAL_INT(-1, ret);
+		TEST_ASSERT_NOT_EQUAL_INT(ENOENT, errno);
+
+		TEST_ASSERT_EQUAL_INT(0, close(ret));
+		TEST_ASSERT_EQUAL_INT(0, remove(PATH));
+		sa.sa_handler = SIG_DFL;
+		TEST_ASSERT_EQUAL(0, sigaction(SIGCHLD, &sa, NULL));
+	}
+}
+
+
+TEST_GROUP(exit);
+
+
+TEST_SETUP(exit)
+{
+}
+
+
+TEST_TEAR_DOWN(exit)
+{
+}
+
+
+TEST(exit, stream_flush)
+{
+	/* Test that exit() force unwritten buffered data to be flushed */
+	pid_t pid;
+	int status;
+
+	FILE *f = fopen(PATH, "w+");
+	TEST_ASSERT_NOT_NULL(f);
+
+	pid = fork();
+	if (pid < 0) {
+		TEST_ASSERT_LESS_THAN_INT(0, pid);
+	}
+	/* child */
+	if (pid == 0) {
+		fprintf(f, TEST_STR);
+		sleep(1); /* Give time  */
+		exit(EXIT_SUCCESS);
+	}
+	/* parent */
+	else {
+		/* Get empty file length */
+		int fd = open(PATH, O_RDWR);
+		TEST_ASSERT_NOT_EQUAL(-1, fd);
+		TEST_ASSERT_EQUAL_INT(0, lseek(fd, 0, SEEK_END));
+
+		wait(&status);
+
+		/* If buffered data was flushed file length should increase */
+		TEST_ASSERT_EQUAL_INT(strlen(TEST_STR), lseek(fd, 0, SEEK_END));
+
+		fclose(f);
+		remove(PATH);
+	}
+}
+
+
+TEST(exit, atexit_few_calls)
+{
+	/* Test that exit() invokes funcs registered by atexit */
+	pid_t pid;
+	int status;
+	fd = open(PATH, O_RDWR | O_CREAT | O_TRUNC);
+	TEST_ASSERT_NOT_EQUAL_INT(-1, fd);
+
+	pid = fork();
+	if (pid < 0) {
+		TEST_ASSERT_LESS_THAN_INT(0, pid);
+	}
+	/* child */
+	if (pid == 0) {
+		TEST_ASSERT_EQUAL_INT(0, atexit(fun1));
+		TEST_ASSERT_EQUAL_INT(0, atexit(fun1));
+		TEST_ASSERT_EQUAL_INT(0, atexit(fun2));
+		exit(EXIT_SUCCESS);
+	}
+	/* parent */
+	else {
+		int ret = wait(&status);
+		TEST_ASSERT_EQUAL_INT(pid, ret);
+		int buf;
+		TEST_ASSERT_EQUAL_INT(0, lseek(fd, 0, SEEK_SET));
+		TEST_ASSERT_EQUAL_INT(sizeof(int), read(fd, &buf, sizeof(int)));
+		TEST_ASSERT_EQUAL_INT(40, buf);
+		TEST_ASSERT_EQUAL_INT(sizeof(int), read(fd, &buf, sizeof(int)));
+		TEST_ASSERT_EQUAL_INT(1, buf);
+		TEST_ASSERT_EQUAL_INT(sizeof(int), read(fd, &buf, sizeof(int)));
+		TEST_ASSERT_EQUAL_INT(2, buf);
+		close(fd);
+		remove(PATH);
+	}
+}
+
+
+TEST(exit, atexit_over32_calls)
+{
+	/* Test that exit() invokes funcs registered by atexit */
+	pid_t pid;
+	int status;
+	int atexit_max = sysconf(_SC_ATEXIT_MAX);
+	fd = open(PATH, O_RDWR | O_CREAT | O_TRUNC);
+	TEST_ASSERT_NOT_EQUAL_INT(-1, fd);
+
+	pid = fork();
+	if (pid < 0) {
+		TEST_ASSERT_LESS_THAN_INT(0, pid);
+	}
+	/* child */
+	if (pid == 0) {
+		int i;
+		for (i = 0; i < atexit_max - 1; i++) {
+			TEST_ASSERT_EQUAL_INT(0, atexit(fun1));
+		}
+		TEST_ASSERT_EQUAL_INT(0, atexit(fun2));
+		TEST_ASSERT_EQUAL_INT(0, atexit(fun1));
```

Comment:
Why is `fun2` used? Can`t we try to register `atexitMax + 1` of `fun1` and assert only `atexitMax` invocations?

Reply:
Last register function is invoked first, so I created fun2 to make distinguish between function calls and ensure they are called in proper order.

Reply:
Is this order asserted somewhere?